common/rate-limiter/src/lib.rs

```diff
@@ -0,0 +1,392 @@
+///
+/// # Terms
+/// ## Key
+/// A `key` is an identifier of the item being rate limited
+///
+/// ## Token
+/// A `token` is the smallest discrete value that we want to rate limit by.  In a situation involving
+/// network requests, this may represent a request or a byte.  `Tokens` are the counters for the
+/// rate limiting, and when there are no `tokens` left in a `bucket`, the `key` is throttled.
+///
+/// ## Bucket
+/// A `bucket` is the tracker of the number of `tokens`.  It has a `bucket size`, and any additional
+/// tokens added to it will "spill" out of the `bucket`.  The `buckets` are filled at an `interval`
+/// with a given `fill rate`.
+///
+/// ## Interval
+/// The `interval` at which we refill *all* of the `buckets` in the token bucket filter. Configured
+/// across the whole token bucket filter.
+///
+/// ## Fill Rate
+/// The rate at which we fill a `bucket` with tokens. Configured per bucket.
+///
+/// ## Bucket Size
+/// Maximum size of a bucket.  A bucket saturates at this size.  Configured per bucket.
+///
+/// # Features
+/// ## Keys
+/// The token bucket takes any key as long as it's hashable.  This should allow it to apply to
+/// many applications that need rate limiters.
+///
+/// ## Bucket sizes and Rates
+/// ### Defaults
+/// There are defaults for bucket size and fill rate, which will apply to unknown keys.
+///
+/// ### Refill Interval
+/// Buckets are refilled automatically at an interval.  To do this synchronously, it calculates the
+/// number of intervals that have passed.  This is done synchronously and in the future may be done
+/// asynchronously.
+///
+pub struct TokenBucketRateLimiter<Key: Eq + Hash + Clone + Debug> {
+    buckets: RwLock<HashMap<Key, Arc<Mutex<Bucket>>>>,
+    default_bucket_size: usize,
+    default_fill_rate: usize,
+    open: bool,
+}
+
+#[allow(dead_code)]
+impl<Key: Eq + Hash + Clone + Debug> TokenBucketRateLimiter<Key> {
+    pub fn new(default_bucket_size: usize, default_fill_rate: usize) -> Self {
+        // TODO: NonZeroUsize?
+        assert!(default_fill_rate > 0, "Fill rate must be greater than 0");
+        assert!(
+            default_bucket_size > 0,
+            "Bucket size must be greater than 0"
+        );
+        assert!(
+            default_bucket_size >= default_fill_rate,
+            "Bucket size must be greater than or equal to fill rate"
+        );
+        Self {
+            buckets: RwLock::new(HashMap::new()),
+            default_bucket_size,
+            default_fill_rate,
+            open: false,
+        }
+    }
+
+    /// Used for testing and to not have a rate limiter
+    pub fn open() -> Self {
+        Self {
+            buckets: RwLock::new(HashMap::new()),
+            default_bucket_size: std::usize::MAX,
+            default_fill_rate: std::usize::MAX,
+            open: true,
+        }
+    }
+
+    /// Retrieve bucket, or create a new one
+    pub fn bucket(&mut self, key: Key) -> Arc<Mutex<Bucket>> {
```

Comment:
if we're using a `RwLock`, then most of these methods should be `&self`. if we're using `&mut self` then there's no point to the `RwLock` since we would have exclusive access to the `TokenBucketRateLimiter` : p